src/Result.php

```diff
@@ -0,0 +1,183 @@
+    {
+        $self = new self();
+        $self->status = self::FOUND;
+        $self->route = $route;
+        $self->handler = $route->handler();
+
+        return $self;
+    }
+
+    public static function createNotFound(): Result
+    {
+        $self = new self();
+        $self->result = [self::NOT_FOUND];
+        $self->status = self::NOT_FOUND;
+
+        return $self;
+    }
+
+    /**
+     * @param string[] $allowedMethods
+     */
+    public static function createMethodNotAllowed(array $allowedMethods): Result
+    {
+        $self = new self();
+        $self->result = [self::METHOD_NOT_ALLOWED, $allowedMethods];
+        $self->status = self::METHOD_NOT_ALLOWED;
+        $self->allowedMethods = $allowedMethods;
+
+        return $self;
+    }
+
+    /**
+     * @param mixed[] $result Result
+     */
+    public static function fromArray(array $result): Result
+    {
+        $self = new self();
+        $self->result = $result;
+        $self->status = $result[0];
+
+        if ($result[0] === self::FOUND) {
+            $self->handler = $result[1];
+            $self->args = $result[2];
+            $self->route = $result[3];
+        }
+
+        return $self;
+    }
+
+    /**
+     * @return mixed
+     */
+    public function handler()
+    {
+        if (! isset($this->result[1])) {
+            return null;
+        }
+
+        return $this->result[1];
+    }
+
+    public function status(): int
+    {
+        return $this->status;
+    }
+
+    /**
+     * @return mixed
+     */
+    public function args()
+    {
+        if (! isset($this->result[2])) {
+            return [];
+        }
+
+        return $this->result[2];
+    }
+
+    public function routeMatched(): bool
+    {
+        return $this->result[0] === self::FOUND;
+    }
+
+    public function methodNotAllowed(): bool
+    {
+        return $this->result[0] === self::METHOD_NOT_ALLOWED;
+    }
+
+    public function routeNotFound(): bool
+    {
+        return $this->result[0] === self::NOT_FOUND;
+    }
+
+    /**
+     * @param mixed $offset
+     */
+    public function offsetExists($offset): bool
+    {
+        return isset($this->result[$offset]);
+    }
+
+    /**
+     * @param mixed $offset
+     *
+     * @return mixed
+     */
+    public function offsetGet($offset)
+    {
+        return $this->result[$offset];
+    }
+
+    /**
+     * @param mixed $offset
+     * @param mixed $value
+     */
+    public function offsetSet($offset, $value): void
+    {
+        throw new RuntimeException(
+            'You can\'t mutate the state of the result'
+        );
+    }
+
+    /**
+     * @param mixed $offset
+     */
+    public function offsetUnset($offset): void
+    {
+        throw new RuntimeException(
+            'You can\'t mutate the state of the result'
+        );
+    }
+
+    /**
+     * Gets the legacy array
+     *
+     * @return mixed[]
+     */
+    public function toArray(): array
+    {
+        return $this->result;
```

Comment:
I believe we should create the array dynamically based on the object state.

Reply:
The array should already have only what is needed. Since the tests still pass I think this is true.

Reply:
Yeah, we're holding something in memory, which is only useful for this method. It doesn't make sense to me.